mautrix-discord: init at unstable-2022-11-04

robintown opened the pull request:
mautrix-discord is a puppeting Discord bridge for Matrix, similar to mx-puppet-discord. Its main advantage over the latter is that mx-puppet-discord is effectively no longer maintained, so the original maintainer has suggested that users move to this bridge instead.

###### Description of changes

This adds a package and NixOS module for mautrix-discord.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

winterqt commented on pkgs/servers/mautrix-discord/default.nix:
```diff
@@ -0,0 +1,26 @@
+
+buildGoModule rec {
+  pname = "mautrix-discord";
+  version = "2022-11-04";
```
```suggestion
  version = "unstable-2022-11-04";
```

winterqt commented on maintainers/maintainer-list.nix:
```diff
@@ -11787,6 +11787,13 @@
     githubId = 521306;
     name = "Rob Glossop";
   };
+  robin = {
+    name = "Robin";
+    email = "robin@robin.town";
+    matrix = "@robin:robin.town";
+    github = "robintown";
+    githubId = 48614497;
+  };
```
Please add this in a separate commit (`maintainers: add robin`).

winterqt commented on nixos/modules/services/matrix/mautrix-discord.nix:
```diff
@@ -0,0 +1,130 @@
+{ config, pkgs, lib, ... }:
```
Please split this addition into a separate commit (`nixos/mautrix-discord: init` or similar).

robintown committed:
maintainers: add robin

robintown committed:
mautrix-discord: init at unstable-2022-11-04

mautrix-discord is a puppeting Discord bridge for Matrix, similar to mx-puppet-discord. Its main advantage over the latter is that mx-puppet-discord is effectively no longer maintained, so the original maintainer has suggested that users move to this bridge instead.

robintown committed:
nixos/mautrix-discord: init

azahi commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -130,6 +130,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - [expressvpn](https://www.expressvpn.com), the CLI client for ExpressVPN. Available as [services.expressvpn](#opt-services.expressvpn.enable).
 
+- [mautrix-discord](https://go.mau.fi/mautrix-discord/) Matrix to Discord hybrid puppeting/relaybot bridge
```
```suggestion
- [mautrix-discord](https://go.mau.fi/mautrix-discord/) Matrix to Discord hybrid puppeting/relaybot bridge.
```
Don't forget to regenerate mdoc after that.

azahi commented on nixos/modules/services/matrix/mautrix-discord.nix:
```diff
@@ -0,0 +1,130 @@
+  settingsFormat = pkgs.formats.yaml {};
+  settingsFile = settingsFormat.generate "mautrix-discord-config.yaml" cfg.settings;
+  runtimeSettingsFile = "${dataDir}/config.yaml";
+
```
```suggestion
```

azahi commented on nixos/modules/services/matrix/mautrix-discord.nix:
```diff
@@ -0,0 +1,130 @@
+  dataDir = "/var/lib/mautrix-discord";
+  registrationFile = "${dataDir}/discord-registration.yaml";
+  cfg = config.services.mautrix-discord;
+  settingsFormat = pkgs.formats.yaml {};
```
```suggestion
  settingsFormat = pkgs.formats.yaml { };
```

azahi commented on nixos/modules/services/matrix/mautrix-discord.nix:
```diff
@@ -0,0 +1,130 @@
+with lib;
+
+let
+  dataDir = "/var/lib/mautrix-discord";
+  registrationFile = "${dataDir}/discord-registration.yaml";
+  cfg = config.services.mautrix-discord;
```
Nit: this should be at the top of the let block.
```suggestion
  cfg = config.services.mautrix-discord;
  dataDir = "/var/lib/mautrix-discord";
  registrationFile = "${dataDir}/discord-registration.yaml";
```

azahi commented on nixos/modules/services/matrix/mautrix-discord.nix:
```diff
@@ -0,0 +1,130 @@
+        umask $old_umask
+      '';
+
+      serviceConfig = {
```
IMO hardening could be expanded a bit.

robintown replied:
The only thing that I could tell was missing compared to similar modules was `ProtectHome`. Is there anything else I should add?

azahi replied:
You can try running `systemd-analyze security <service-name>` and fixing up some obvious stuff. Here's an [example](https://github.com/NixOS/nixpkgs/blob/733dcaf53df671a4f58a6d1fb76cc6c476568bf2/nixos/modules/services/security/endlessh-go.nix#L98-L129). It doesn't have to be this excessive, though.

azahi commented on pkgs/servers/mautrix-discord/default.nix:
```diff
@@ -0,0 +1,26 @@
+    owner = "mautrix";
+    repo = "discord";
+    rev = "f53975cc91e3c643a722adf0d3e0dfb98d0127a2";
+    sha256 = "sha256-xUALcN5oQfwC6gmOeygCkOAXJrJzNitBxHRFAKKgFvE=";
```
```suggestion
    hash = "sha256-xUALcN5oQfwC6gmOeygCkOAXJrJzNitBxHRFAKKgFvE=";
```

azahi commented on pkgs/servers/mautrix-discord/default.nix:
```diff
@@ -0,0 +1,26 @@
+
+  buildInputs = [ olm ];
+
+  vendorSha256 = "sha256-yday2mSnPwuhXWkCG4XY7qoBl3DXHcSvzBoZbjgYz/c=";
```
```suggestion
  vendorHash = "sha256-yday2mSnPwuhXWkCG4XY7qoBl3DXHcSvzBoZbjgYz/c=";
```

azahi commented on pkgs/servers/mautrix-discord/default.nix:
```diff
@@ -0,0 +1,26 @@
+  buildInputs = [ olm ];
+
+  vendorSha256 = "sha256-yday2mSnPwuhXWkCG4XY7qoBl3DXHcSvzBoZbjgYz/c=";
+
```
Ref: #177698
```suggestion
ldflags = [ "-s" "-w" ];

```

azahi reviewed (changes requested):
The module also lacks a test, we probably should add at least a simple HTTP GET healthcheck.
